Objects/unionobject.c

```diff
@@ -468,23 +547,12 @@ make_union(PyObject *args)
 {
     assert(PyTuple_CheckExact(args));
 
-    args = dedup_and_flatten_args(args);
```

Comment:
We still need this logic for the callsite in `union_getitem`.

Consider this case:

```pycon
>>> from typing import TypeVar
>>> T = TypeVar("T")
>>> (list[T] | list[int])
list[~T] | list[int]
>>> (list[T] | list[int])[int]
list[int]
```

If I'm reading your code correctly, it would no longer deduplicate the two members. It would be good to also add a test case based on this example.

We should probably put the deduping logic directly in `union_getitem`.

Reply:
@JelleZijlstra Actually, we don't need it, as far as such case already handled by tests:
https://github.com/python/cpython/blob/1cd8c29dace2dc6b91503803113fea4288ca842b/Lib/test/test_types.py#L823

That's because after parameter substitution new `UnionType` is recreated by reducing `newargs` using `bitwise or` operator:

https://github.com/python/cpython/blob/1cd8c29dace2dc6b91503803113fea4288ca842b/Objects/unionobject.c#L380-L388